sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Cast.scala

```diff
@@ -316,6 +316,15 @@ abstract class CastBase extends UnaryExpression with TimeZoneAwareExpression wit
   override lazy val resolved: Boolean =
     childrenResolved && checkInputDataTypes().isSuccess && (!needsTimeZone || timeZoneId.isDefined)
 
+  override lazy val roughlyCanonicalized: Expression = {
```

Comment:
Should we maybe consider having a separate trait for this?

It looks to me that there are only specific expressions that can be optimized by roughly canonicalizing .. e.g.) we can have a trait, and call `roughlyCanonicalized` if it inherits the trait at `canonicalized`.